use animation frames for memory autoplay

uhthomas opened the pull request:
The current implementation mixes intervals and animation frames, which is a little convoluted. The use of intervals means that the animation is not going to be smooth and may have strange behaviour when the window is moved to the background. It's possible that the current animation frames could pile up and run all at once which would be undesirable.

Moving everything into animation frames means the code is simpler and easier to reason about. It should also be more performant and less buggy.

uhthomas commented:
I've tested and it works pretty well, but resumption seems broken. I guess I should fix that.

jrasm91 commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -42,119 +45,108 @@
+	});
 
-		currentMemory = $memoryStore[currentIndex];
+	onDestroy(async () => {
+		if (browser) pause();
+	});
```
```suggestion
	onDestroy(() => {
		if (browser) {
		  pause();
		}  
	});
```

jrasm91 commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -42,119 +45,108 @@
 
-		nextIndex = currentIndex + 1;
-		nextMemory = $memoryStore[nextIndex];
+	const canAdvance = () => !!(nextMemory || nextAsset);
```
```suggestion
	$: canAdvance = !!(nextMemory || nextAsset);
```

jrasm91 commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -246,29 +239,33 @@
-										backgroundColor="#202123"
-										on:click={toNextCurrentAsset}
-									/>
+									{#if nextMemory || assetIndex < currentMemory.assets.length - 1}
```
I think you can use `nextMemory || nextAsset` here.

jrasm91 commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -246,29 +239,33 @@
-										backgroundColor="#202123"
-										on:click={toPreviousCurrentAsset}
-									/>
+									{#if previousMemory || assetIndex > 0}
```
I think you can use `previousMemory || previousAsset` here

jrasm91 commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -246,29 +239,33 @@
+										<CircleIconButton
+											logo={ChevronLeft}
+											backgroundColor="#202123"
+											on:click={toPreviousAsset}
```
Usually click methods use `handlePreviousAsset` or `onPreviousAsset` naming convention, depending if it is being handled or propagated.

uhthomas replied:
Makes sense, this is just what the original code was.

jrasm91 commented on web/src/lib/components/memory-page/memory-viewer.svelte:
```diff
@@ -42,119 +45,108 @@
+			}
+		};
+
+		const requestDraw = () => (animationFrameRequest = requestAnimationFrame(draw));
```
It is a little strange to define this function here, reference it above (before the definition), and also it's implementation references `draw` which also calls it. I think it'd be possible to actually have the entire `draw` method be defined outside of the `play` method, and `play` only starts the animationFrameRequest loop if necessary.

uhthomas replied:
Yeah, it is now. There was a point where I was working on it where `start` was defined in `play`. That's been moved out to handle pausing and navigation correctly, so I can now separate the function too.

jrasm91 reviewed (commented):
We don't have one line if statements anywhere else in the codebase. In general, I think most of the developers (at least myself) prefer _not_ using if/for on single lines without braces.

Looks much cleaner though, thanks for adding those changes in. I think the next/current asset stuff as well.

uhthomas committed:
use animation frames for memory autoplay

The current implementation mixes intervals and animation frames, which is a
little convoluted. The use of intervals means that the animation is not going
to be smooth and may have strange behaviour when the window is moved to the
background. It's possible that the current animation frames could pile up and
run all at once which would be undesirable.

Moving everything into animation frames means the code is simpler and easier to
reason about. It should also be more performant and less buggy.

alextran1502 committed:
merge main
